File: frontend/src/components/benchmark/content/triangle_detection/implementations/triangle_detection.js
class TriangleDetection {
    constructor() {
        this.adjacencyList = new Map();
    }

    addEdge(u, v) {
        if (!this.adjacencyList.has(u)) {
            this.adjacencyList.set(u, new Set());
        }
        if (!this.adjacencyList.has(v)) {
            this.adjacencyList.set(v, new Set());
        }
        this.adjacencyList.get(u).add(v);
        this.adjacencyList.get(v).add(u);
    }

    buildFromEdges(edges) {
        for (const [u, v] of edges) {
            this.addEdge(u, v);
        }
    }

    countTriangles() {
        let count = 0;
        for (const [u, neighbors] of this.adjacencyList) {
            for (const v of neighbors) {
                if (v > u) {
                    for (const w of neighbors) {
                        if (w > v && this.adjacencyList.get(v).has(w)) {
                            count++;
                        }
                    }
                }
            }
        }
        return count;
    }

    findTriangles() {
        const triangles = [];
        for (const [u, neighbors] of this.adjacencyList) {
            for (const v of neighbors) {
                if (v > u) {
                    for (const w of neighbors) {
                        if (w > v && this.adjacencyList.get(v).has(w)) {
                            triangles.push([u, v, w]);
                        }
                    }
                }
            }
        }
        return triangles;
    }

    hasTriangle() {
        for (const [u, neighbors] of this.adjacencyList) {
            for (const v of neighbors) {
                if (v > u) {
                    for (const w of neighbors) {
                        if (w > v && this.adjacencyList.get(v).has(w)) {
                            return true;
                        }
                    }
                }
            }
        }
        return false;
    }
}

module.exports = TriangleDetection;

File: frontend/src/components/benchmark/content/triangle_detection/implementations/triangle_detection.rs
use std::collections::{HashMap, HashSet};

pub struct TriangleDetection {
    adjacency_list: HashMap<i32, HashSet<i32>>,
}

impl TriangleDetection {
    pub fn new() -> Self {
        TriangleDetection {
            adjacency_list: HashMap::new(),
        }
    }

    pub fn add_edge(&mut self, u: i32, v: i32) {
        self.adjacency_list.entry(u).or_insert_with(HashSet::new).insert(v);
        self.adjacency_list.entry(v).or_insert_with(HashSet::new).insert(u);
    }

    pub fn build_from_edges(&mut self, edges: &[(i32, i32)]) {
        for &(u, v) in edges {
            self.add_edge(u, v);
        }
    }

    pub fn count_triangles(&self) -> usize {
        let mut count = 0;
        for (&u, neighbors) in &self.adjacency_list {
            for &v in neighbors {
                if v > u {
                    for &w in neighbors {
                        if w > v {
                            if let Some(neighbors_v) = self.adjacency_list.get(&v) {
                                if neighbors_v.contains(&w) {
                                    count += 1;
                                }
                            }
                        }
                    }
                }
            }
        }
        count
    }

    pub fn find_triangles(&self) -> Vec<(i32, i32, i32)> {
        let mut triangles = Vec::new();
        for (&u, neighbors) in &self.adjacency_list {
            for &v in neighbors {
                if v > u {
                    for &w in neighbors {
                        if w > v {
                            if let Some(neighbors_v) = self.adjacency_list.get(&v) {
                                if neighbors_v.contains(&w) {
                                    triangles.push((u, v, w));
                                }
                            }
                        }
                    }
                }
            }
        }
        triangles
    }

    pub fn has_triangle(&self) -> bool {
        for (&u, neighbors) in &self.adjacency_list {
            for &v in neighbors {
                if v > u {
                    for &w in neighbors {
                        if w > v {
                            if let Some(neighbors_v) = self.adjacency_list.get(&v) {
                                if neighbors_v.contains(&w) {
                                    return true;
                                }
                            }
                        }
                    }
                }
            }
        }
        false
    }
}